src/satosa/frontends/saml2.py

```diff
@@ -299,18 +300,36 @@ def _handle_authn_response(self, context, internal_response, idp):
             attributes_to_remove = custom_release.get("exclude", [])
             for k in attributes_to_remove:
                 ava.pop(k, None)
-
-        name_id = NameID(text=internal_response.user_id,
-                         format=hash_type_to_saml_name_id_format(
-                             internal_response.user_id_hash_type),
-                         sp_name_qualifier=None,
-                         name_qualifier=None)
+        idp_conf = self.idp_config.get('service', {}).get('idp', {})
+        try:
+            config_nameidformat = idp_conf['name_id_format']
+            if isinstance(config_nameidformat, list):
+                config_nameidformat = config_nameidformat[0]
+        except KeyError:
+            config_nameidformat = None
+            satosa_logging(logger, logging.DEBUG, "No nameid format configured for frontend idp",
+                           context.state)
+        if config_nameidformat == NAMEID_FORMAT_EMAILADDRESS:
+            nameidfmt = NAMEID_FORMAT_EMAILADDRESS
+            nameidval = internal_response.attributes.get('mail', None)[0]
```

Comment:
This will fail if `'mail'` is not set as `None` will be returned and tried to be accessed as an array (`...[0]`). It should fallback to `[None]`.

```python
nameidval = internal_response.attributes.get('mail', [None])[0]
```